access tasks configs as preferences

elaihau opened the pull request:
- In current Theia, the task extension is responsible for reading,
writing, and watching `tasks.json` files. With this change, the task
extension does not access `tasks.json` files directly, and leaves the
work to the preference extension.
- resolves #5013


#### How to test
There are no changes in terms of functionalities. Therefore, to verify this change, we would need to go through the use cases that are already supported by Theia, and  check if they are still properly functional. Those use cases include but not limit to:

As a Theia user I am able to

- run tasks defined in `tasks.json`
- run detected tasks contributed by Theia plugins
- run detected tasks contributed by Theia extensions (Please note, this is broken on master branch and reported in https://github.com/eclipse-theia/theia/issues/6204)
- configure tasks defined in `tasks.json`
- configure detected tasks (Please note, "configuring tasks" by clicking the gear icons in tasks' quick open menu is broken on master branch and reported in https://github.com/eclipse-theia/theia/issues/6212. In order to test this feature we would have to use the top menu bar `Terminal` -> `Configure Tasks ...`)
- see the recently used tasks be grouped together and show up at the top of the tasks' quick open menu
- see the "detected tasks" displayed as "configured tasks", if they are configured / customized in the `tasks.json`
- see the problem matcher items show up in the prompt, if a task that is not associated with any problem matchers is started
   - and see the selected problem matcher item written into `tasks.json` 
- do all the items described above in both single root and multi root workspace.

akosyakov commented on packages/task/src/browser/task-configuration-model.ts:
```diff
@@ -0,0 +1,93 @@
+        return this.json.configurations;
+    }
+
+    reconcile(): void {
```
`protected`

akosyakov commented on packages/task/src/browser/task-configuration-manager.ts:
```diff
@@ -0,0 +1,200 @@
+            toDelete.delete(key);
+            if (!this.models.has(key)) {
+                const model = new TaskConfigurationModel(key, this.preferences);
+                model.onDidChange(() => this.onChangedTaskConfigEmitter.fire({ uri: key, type: FileChangeType.ADDED }));
```
Should not it be `UPDATED`?

elaihau replied:
i am not sure which (`ADDED` vs `UPDATED`) is more appropriate. to be honest i feel neither is 100% correct :)

when a new project folder is added to the workspace, from the taskManager's perspective, task configs from that new project folder are "ADDED" to the workspace. However, from the workspace's perspective, it is an `UPDATE` to the pool of task configs. 

i am ok with changing `ADDED` to `UPDATED` - it has no impact to the end effect, because subscriber of this event only cares about "delete" vs "not-a-delete".

akosyakov replied:
it also strange that we use `FileChangeEvent` here, then notion of files is lost already behind the preference service. Maybe we need another event? `TaskConfigChange` which has only an uri and then `TaskConfiguraionManager` has a method to check whether a config exist for such uri?

akosyakov commented on packages/task/src/browser/task-configuration-manager.ts:
```diff
@@ -0,0 +1,200 @@
+                const model = new TaskConfigurationModel(key, this.preferences);
+                model.onDidChange(() => this.onChangedTaskConfigEmitter.fire({ uri: key, type: FileChangeType.ADDED }));
+                model.onDispose(() => this.models.delete(key));
+                this.models.set(key, model);
```
Should not we fire an added event here? The first event happens in the constructor, `onDidChange` is registered too late for it.

akosyakov commented on packages/task/src/browser/task-configuration-manager.ts:
```diff
@@ -0,0 +1,200 @@
+    @inject(WorkspaceVariableContribution)
+    protected readonly workspaceVariables: WorkspaceVariableContribution;
+
+    protected readonly onDidChangeEmitter = new Emitter<void>();
```
never used

akosyakov commented on packages/task/src/browser/task-configurations.ts:
```diff
@@ -18,16 +18,12 @@ import { inject, injectable, postConstruct } from 'inversify';
 import { WorkspaceService } from '@theia/workspace/lib/browser';
-import { open, OpenerService } from '@theia/core/lib/browser';
-import { Resource } from '@theia/core';
+import { /*open,*/ OpenerService } from '@theia/core/lib/browser';
```
leftover?

akosyakov commented on packages/task/src/browser/task-configurations.ts:
```diff
@@ -85,41 +78,32 @@ export class TaskConfigurations implements Disposable {
     @postConstruct()
     protected init(): void {
+        this.toDispose.push(
+            this.taskConfigurationManager.onChangedTaskConfig(async change => {
```
fyi: `onDidChangeTaskConfig` or `onTaskConfigChanged`, the former is according to our coding guidelines

akosyakov commented on packages/task/src/browser/task-configurations.ts:
```diff
@@ -85,41 +78,32 @@ export class TaskConfigurations implements Disposable {
+        this.toDispose.push(
+            this.taskConfigurationManager.onChangedTaskConfig(async change => {
+                try {
+                    await this.onDidTaskFileChange([change]);
```
looking at the implementation, strange that `removeTasks` is exposed, it should be called by clients since it does not change underlying files

the same about `removeTaskCustomizations`

elaihau replied:
i will change them to `private`. Thank you !

akosyakov commented on packages/task/src/browser/task-configurations.ts:
```diff
@@ -365,14 +277,13 @@ export class TaskConfigurations implements Disposable {
 
         try {
-            await open(this.openerService, new URI(configFileUri));
+            this.taskConfigurationManager.openConfiguration(sourceFolderUri);
```
`await` otherwise `try-catch` does not make sense

akosyakov commented:
@RomanNikitenko @vince-fugnitto if you do smoke testing it would help too, just to be safe, i.e. with some VS Code extensions.

elaihau commented:
Thank you for the review Anton !
I addressed the comments, and will test if task configs can be defined in preferences. Never thought of trying with this change :)

akosyakov commented:
@elaihau more important that it can be loaded from `.vscode`, from settings.json is probably very rare

akosyakov commented on packages/task/src/browser/task-configuration-manager.ts:
```diff
@@ -0,0 +1,198 @@
+                model.onDispose(() => this.models.delete(key));
+                this.models.set(key, model);
+            }
+            this.onDidChangeTaskConfigEmitter.fire({ uri: key, type: FileChangeType.UPDATED });
```
It should happen only in `if` branch

elaihau commented:
@vince-fugnitto thank you for the review, really apprecited !

I tested the scenario where task configs are defined in `.vscode/tasks.json`

![Peek 2019-09-28 12-23](https://user-images.githubusercontent.com/37082801/65819487-1bb63180-e1eb-11e9-9155-562f81ac4eeb.gif)

vince-fugnitto commented:
> @vince-fugnitto thank you for the review, really apprecited !
> I tested the scenario where task configs are defined in `.vscode/tasks.json`

I confirmed that it also works for me to define tasks under `.vscode/tasks.json`.
I however attempted to have both `.vscode` and `.theia` tasks and could not (only sees the `.theia` tasks). This is likely to be the expected behavior but I was not sure.

elaihau commented:
> I confirmed that it also works for me to define tasks under `.vscode/tasks.json`.
> I however attempted to have both `.vscode` and `.theia` tasks and could not (only sees the `.theia` tasks). This is likely to be the expected behavior but I was not sure.

yep i believe what you described was from Anton's code: https://github.com/eclipse-theia/theia/blob/02c5c1d4539efa8b19f9d0991be9827c88e184f6/packages/core/src/browser/preferences/preference-configurations.ts#L37 and I think it's expected.
